server/src-lib/Hasura/Server/Telemetry/Counters.hs

```diff
@@ -0,0 +1,183 @@
+{-# LANGUAGE DuplicateRecordFields #-}
+{-# LANGUAGE RecordWildCards       #-}
+{-|
+  Counters used in telemetry collection. Additional counters can be added here.and
+  serviced in "Hasura.Server.Telemetry".
+-}
+
+module Hasura.Server.Telemetry.Counters
+  (
+  -- * Service timing and counts, by various dimensions
+  -- ** Local metric recording
+    recordTimingMetric
+  , RequestDimensions(..), RequestTimings(..)
+  -- *** Dimensions
+  , CacheHit(..), QueryType(..), Locality(..), Transport(..)
+  -- ** Metric upload
+  , dumpServiceTimingMetrics
+  , ServiceTimingMetrics(..)
+  , ServiceTimingMetric(..)
+  , RunningTimeBucket(..)
+  , RequestTimingsCount(..)
+  )
+  where
+
+import qualified Data.Aeson            as A
+import qualified Data.Aeson.Casing     as A
+import qualified Data.Aeson.TH         as A
+import           Data.Hashable
+import qualified Data.HashMap.Strict   as HM
+import           Data.IORef
+import           Data.Time.Clock.POSIX (POSIXTime, getPOSIXTime)
+import           GHC.IO.Unsafe         (unsafePerformIO)
+import           Hasura.Prelude
+
+-- | The properties that characterize this request. The dimensions over which
+-- we collect metrics for each serviced request.
+data RequestDimensions =
+  RequestDimensions {
+        telemCacheHit  :: !CacheHit
+      , telemQueryType :: !QueryType
+      , telemLocality  :: !Locality
+      , telemTransport :: !Transport
+      }
+      deriving (Show, Generic, Eq)
+
+instance Hashable RequestDimensions
+
+
+-- | Accumulated time metrics.
+data RequestTimings  =
+  RequestTimings {
+      telemTimeIO  :: !Seconds
+    -- ^ Time spent waiting on PG/remote http calls
+    , telemTimeTot :: !Seconds
+    -- ^ Total service time for request (including 'telemTimeIO')
+    }
+
+-- | Sum
+instance Semigroup RequestTimings where
+  RequestTimings a b <> RequestTimings x y = RequestTimings (a+x) (b+y)
+
+-- | 'RequestTimings' along with the count
+data RequestTimingsCount  =
+  RequestTimingsCount {
```

Comment:
The current model does not seem to enable the ability to calculate percentiles, which are extremely useful for performance tracking. Would you guys be open to explore an alternative design that would enable calculations beyond totals and averages?

Reply:
For now for the purposes of telemetry we're just collecting counts and timings for a few hand chosen buckets, which should be fine for our purposes, rather than using hdrhistogram or something similar. But I agree with you that percentiles are useful!